Fix rwdb memory leak with online_delete and remove flatmap

sublimator opened the pull request.

dangell7 committed:
fix online delete

sublimator committed:
remove flatmap database implementation

sublimator committed:
[wip] add debugging

dangell7 committed:
add test file

sublimator committed:
fix: add automatic LEDGER_HISTORY cleanup to RWDB

- Extract deletion logic into lock-free private helpers to avoid deadlock
- Add automatic cleanup in saveValidatedLedger when LEDGER_HISTORY is set
- Clear transactions, account transactions, and ledgers older than cutoff

sublimator committed:
Revert "fix: add automatic LEDGER_HISTORY cleanup to RWDB"

This reverts commit e4a6def5b545723ce9aea4b91a1b7c3a23b3e096.

sublimator committed:
fix: enforce online_delete requirement for RWDB backend

RWDB stores data in memory and will grow unbounded without online_delete
since NodeStore has no delete methods. Add startup check to prevent OOM.
Exception for standalone mode (tests). Update config documentation to
clarify RWDB requires online_delete while it's optional for other backends.

sublimator committed:
fix: path is NOT required for rwdb

sublimator committed:
chore: tidy up and format

sublimator committed:
feat: add RWDB online_delete enforcement with test override mechanism

- RWDB now requires online_delete configuration to prevent OOM
- Exception: standalone mode (tests) bypasses this requirement
- Added _online_delete_standalone_override config for testing enforcement
- Tests can now validate the enforcement logic by overriding standalone detection
- Added comprehensive tests covering all enforcement scenarios

sublimator committed:
feat: enforce online_delete requirement for RWDB to prevent OOM

RWDB (in-memory backend) now requires online_delete configuration to prevent
unbounded memory growth. Exception: standalone mode (used by tests) bypasses
this requirement since tests run for short durations.

- Add enforcement check in SHAMapStoreImp constructor
- Tests use Config::setupControl() to simulate non-standalone environment
- Comprehensive test coverage for all enforcement scenarios

sublimator commented on src/test/app/SHAMapStore_test.cpp:
```diff
@@ -517,12 +521,96 @@ class SHAMapStore_test : public beast::unit_test::suite
+    void
+    testRWDBOnlineDeleteEnforcement()
+    {
+        testcase("RWDB online_delete enforcement");
```
I don't really like this here, but it is what it is 
It's testing code in SHAMapStoreImp.cpp, which had preexisting validation code that throwed exceptions
Take the L and move on?

sublimator commented on src/test/app/SHAMapStore_test.cpp:
```diff
@@ -517,12 +521,96 @@ class SHAMapStore_test : public beast::unit_test::suite
+
+        // Test 2: RWDB without online_delete NOT in standalone mode (should
+        // throw)
+        try
```
If the component instantiation order changes, these tests could start failing. They are a bit iffy/brittle

sublimator commented on src/ripple/app/misc/SHAMapStoreImp.cpp:
```diff
@@ -118,6 +118,19 @@ SHAMapStoreImp::SHAMapStoreImp(
     get_if_exists(section, "online_delete", deleteInterval_);
 
+    // RWDB (in-memory backend) requires online_delete to prevent OOM
+    // Exception: standalone mode (used by tests) doesn't need online_delete
```
todo, move to ripple::Config::loadFromString

sublimator committed:
refactor: move RWDB online_delete validation to Config::loadFromString

Move validation from SHAMapStoreImp constructor to configuration loading phase
as indicated by GitHub review comment. This improves architecture by catching
configuration errors earlier in the initialization process.

- Remove duplicate validation from SHAMapStoreImp.cpp
- Keep validation only in Config::loadFromString()
- Move tests from SHAMapStore_test to Config_test
- Clean up test string formatting to use regular literals

RichardAH commented on src/ripple/app/rdb/backend/RWDBDatabase.h:
```diff
@@ -463,11 +457,9 @@ class RWDBDatabase : public SQLiteDatabase
             {
-                size += sizeof(uint32_t) +
-                    innerMap.size() * (sizeof(uint32_t) + sizeof(size_t));
+                size += sizeof(uint32_t) + txVector.size() * sizeof(AccountTx);
```
is this right?
`    using AccountTx =
        std::pair<std::shared_ptr<Transaction>, std::shared_ptr<TxMeta>>;`
shouldn't we iterate txVector and find the byte sizes of the txns and meta?

same above for the size of ledgers

RichardAH commented on src/test/rdb/RelationalDatabase_test.cpp:
```diff
@@ -0,0 +1,632 @@
+    }
+
+    void
+    run() override
```
should these tests be run against multiple relational database backends?

sublimator committed:
fix: add actual transaction data to RWDB memory usage calculations

The existing memory usage functions only counted structural overhead
(sizeof of containers and pointers) but missed the actual transaction
and metadata blob sizes. This caused severe underreporting - showing
KBs when actually using MBs or GBs.

Changes:
- Keep existing structural overhead calculations
- Add actual transaction/metadata serialized data sizes
- Use transactionMap_ as single source to avoid double-counting
- Add MAP_NODE_OVERHEAD constant for red-black tree nodes (~40 bytes each)
- Use vector::capacity() instead of size() for actual allocated memory
- Include ledger's transaction map node overhead in ledger calculations
- Change internal calculation to uint64_t to prevent overflow
- Add clear comments explaining what each section measures

These improvements provide much more accurate memory reporting for
monitoring and diagnostic purposes.

sublimator committed:
test: parameterize RelationalDatabase_test for multiple backends

- Add backend parameter to all test methods
- Parse backends from --unittest-arg (CSV format: sqlite,rwdb)
- Skip dbHasSpace checks for SQLite (uses in-memory databases in standalone mode)
- Remove unnecessary database path setup/cleanup
- Rename getDB() to getInterface()

sublimator committed:
test: relax memory size assertions in RelationalDatabase_test

- Replace exact KB values with reasonable ranges
- RWDB: Check starts at 0, increases with data, maintains relationships
- SQLite: Check non-negative values and non-decreasing behavior
- Both: Focus on testing relationships rather than brittle exact values

sublimator committed:
refactor: simplify getBackends() with early return for default case

sublimator committed:
chore: clean up after you know who

sublimator committed:
docs: clarify KB rounding behavior in RWDB memory assertions
